elixir-phoenix-deployement.md

```diff
@@ -0,0 +1,162 @@
+
+We are used to deploy Elixir/Phoenix on Heroku
+and we have already a ["how to"](https://github.com/dwyl/learn-phoenix-framework/blob/master/heroku-deployment.md) guide of this process.
+
+However there are some limitations when using Heroku with Phoenix (see https://hexdocs.pm/phoenix/heroku.html):
+![image](https://user-images.githubusercontent.com/6057298/83642784-2e1d1200-a5a7-11ea-8c97-9c7dd920469c.png)
+
+pricing:
+![image](https://user-images.githubusercontent.com/6057298/83641712-cc0fdd00-a5a5-11ea-8cbf-49981ae3747e.png)
+
+The pricing above doesn't include the database which needs to be added to the total cost, see https://elements.heroku.com/addons/heroku-postgresql
+
+
+### Gigalixir
+
+Similar to Heroku [Gigalixir](https://www.gigalixir.com/) provides
+a platform which focuses _exclusively_ on deploying Elixir/Phoenix application.
+Having an exclusive focus on deploying Elixir Apps gives Gigalixir several key advantages:
+1. Zero-downtime blue-green continuous deployment.
+see: https://martinfowler.com/bliki/BlueGreenDeployment.html
+2. No limit to concurrent connections.
+3. All clustering handled transparently. 
+(so if you need to scale your app beyond 10k concurrent users, 
+you don't have to pay exponentially more for bigger "dynos" 
+the way you are forced to scale vertically on Heroku)
+
+Deploying: https://elixircasts.io/deploying-with-gigalixir-%28revised%29
+
+Pricing:
+
+![image](https://user-images.githubusercontent.com/6057298/83643629-290c9280-a5a8-11ea-89da-648daec204f3.png)
+
+Gigalixir vs Heroku:
+
+![image](https://user-images.githubusercontent.com/6057298/83644108-d1225b80-a5a8-11ea-869a-c8c0e4a28012.png)
+
+Like Heroku you need to add the cost for using the database:
+
+![image](https://user-images.githubusercontent.com/6057298/83644296-09c23500-a5a9-11ea-9fdd-c50532195c95.png)
+
+see also tiers pricing page: https://gigalixir.readthedocs.io/en/latest/tiers-pricing.html
+
+
+### Render
+
+[Render](https://render.com/) is another Paas similar to Heroku
+
+pricing:
+
+![image](https://user-images.githubusercontent.com/6057298/83645052-e9df4100-a5a9-11ea-8d3e-f1b18bb4cc02.png)
+
+
+
+## Virtual Private Server
+
+VPS allow us to manage ourself the deployement
+setup. This allow us to customise the server and
+the costs linked to it.
+
+
+### Linode
+
+#### Ubuntu
+
+Linode provides and support Ubuntu:
+![image](https://user-images.githubusercontent.com/6057298/83647811-3bd59600-a5ad-11ea-893c-b99e3df7f605.png)
+
+
+From there the idea is install Erlang and Elixir on the server
+and then to run the application.
+
+- Install Erlang/Elixir with asdf:
+    - `git clone https://github.com/asdf-vm/asdf.git ~/.asdf --branch v0.7.8`
+    - Edit `~/.bashrc file` and add `. $HOME/.asdf/asdf.sh` and run `source ~/.barhrc` to access the `asdf` command
+    - Install required pacakges for Erlang `sudo apt install libssl-dev make automake autoconf libncurses5-dev gcc`
+    - Add erlang plugin to asdf: `asdf plugin-add erlang`
```

Comment:
If we are using Linux as our Host OS then we 100% do _not_ need to install Erlang/Elixir on the VM. 
We can use Travis-CI which runs Ubuntu 18.04 ("Bionic") https://docs.travis-ci.com/user/reference/bionic
to build the package using [`distiliery`](https://github.com/bitwalker/distillery) and deploy it to the VM using [`edeliver`](https://github.com/edeliver/edeliver). 

> And as soon as Ubuntu `20.04 Focal` becomes available on Travis-CI  we can upgrade! Right now only `18.04` is supported as noted in: https://github.com/internetarchive/openlibrary/pull/3407#issuecomment-638637187
I fear that I may not have _fully_ explained "Continuous Integration + Continuous Deployment" in the issue ... #59  💭 
Perhaps I should have been more explicit that we would be following the "12 Factor App" best practices and separating the "Build Phase" from the "Run Phase" see: https://12factor.net/build-release-run
This strict separation ensures that there is no way to _modify_ the code on the production VM which _dramatically_ enhances security because it almost eliminates the risk of [arbitrary code execution](https://en.wikipedia.org/wiki/Arbitrary_code_execution) (_at least from the Erlang/Elixir perspective_).

The whole point is that our Continuous Integration server (Travis-CI) does the testing ***and*** build phase and then just _ships_ an **`executable binary`** of the application to the Linode VM. We _definitely_ don't want to be installing (_and thus having to **manage/maintain**_) **`asdf`** then `erlang` and `elixir` on the VM if we don't absolutely _need_ to.

We already demonstrated how to build and deploy a Phoenix App to a VM using [`distillery`](https://github.com/bitwalker/distillery) and [`edeliver`](https://github.com/edeliver/edeliver):
https://github.com/dwyl/learn-microsoft-azure#deploy-your-phoenix-web-app-using-edeliver
When we shipped [Healthlocker](https://github.com/healthlocker/healthlocker) to Azure in **2017**. 
See: https://github.com/healthlocker/healthlocker/blob/master/.deliver/config and: [`.travis.yml#L33-L35`](https://github.com/healthlocker/healthlocker/blob/becff6feafb3e10ebe061f7056bcc04bf4755c0c/.travis.yml#L33-L35)
This worked without any issues and we spent a _lot_ of time documenting all the scripts. e.g [`.deliver/version.sh`](https://github.com/healthlocker/healthlocker/blob/becff6feafb3e10ebe061f7056bcc04bf4755c0c/.deliver/version.sh)

The only reason I did not _specify_ using `edeliver` in the issue was because I wanted there to be a "fresh eyes" investigation into the _best_ way of doing Phoenix deployment to a VPS in 2020. _Installing_ the Erlang on the VPS and building the project on the production server is _definitely_ not the way we want to do it. The "_right_" way of doing this is to build the release (_binary_) on Travis-CI (_using someone else's compute time that they have generously provided for free!_) and uploading it to S3 so that we have a full history of all our release objects. Then deploy the release to the VPS which does _not_ have Erlang or _anything_ unnecessary installed on it. By _minimising_ the number of programs installed on the VM we minimise _maintenance_ and simplify our lives. The moment we introduce a new tool (e.g. `asdf`) we need to _understand_ the security+maintenance implications. If the `asdf` tool is installing any dependencies or running arbitrary shell scripts, who is _responsible_ for ensuring that it's not doing something undesirable (malicious) like opening a TCP port or leaking credentials? Where is `asdf` sourcing it's binaries for `erlang` and `elixir` from? Who has control over them? Is there a security audit? _Who_ is [Trevor Brown](https://github.com/Stratus3D) and _why_ should I trust them? He _seems_ like a "nice guy" https://www.linkedin.com/in/trevor-brown-11136938/ but how do we _know_ he is following security best practices cannot be ["spearphished"](https://en.wikipedia.org/wiki/Phishing#Spear_phishing), pwned and infect all the servers running `asdf`?

P.S. sorry to be a pain about this in the in the PR review, I should have made it clearer in the issue.
12 Factor App. Don't install build dependencies on the Production Server. 👍